src/scheduler/scheduler.rs

```diff
@@ -251,6 +225,7 @@ impl<VM: VMBinding> GCWorkScheduler<VM> {
 
     fn are_buckets_drained(&self, buckets: &[WorkBucketStage]) -> bool {
         buckets.iter().all(|&b| self.work_buckets[b].is_drained())
+            && self.pending_messages.load(Ordering::Relaxed) == 0
```

Comment:
`pending_messages` is stored with `Ordering::SeqCst` in `controller.rs`, but is loaded with `Ordering::Relaxed` here, and I can't find any fences (`std::sync::atomic::fence`).  Is there any reason to use `Relaxed` instead of `SeqCst`?  In theory, if this load has `Relaxed` order, subsequent code in this worker thread may not see the effect of store operations in the work packet (such as StopMutators) executed by the coordinator thread, due to reordering in both machine and compiler level.

Reply:
I'm also not 100% sure about this. We may need SeqCst so that the stores to `pending_messages` will be seen immediately. But currently, it is safe if the load does not see the new value. In this case, no buckets will open and the workers will be put to sleep. Our only use case for the coordinator worker is to execute the "StopMutators" packets. The packet will wake up the workers anyway, and the workers will decide to open more buckets again (or execute new packets).  Probably I'll just change this to SeqCst.

Reply:
It is not about this load not seeing `pending_messages` being updated or not opening buckets.  It is the opposite:

**If I see `pending_messages == 0`**, can I guarantee that I see the effect of `StopMutators`?

In other words, if

1.  The coordinator puts a work packet in the `Prepare` bucket, and 
2.  The coordinator sets `pending_messages` to 0, and
3.  A worker loads from `pending_messages`, and sees 0, and
4.  The worker looks at the `Prepare` bucket,

Is the worker guaranteed that it sees the length of the `Prepare` bucket to be at least one (seeing the packet added by the coordinator)?  It is guaranteed if (4) happens after (1).  We know that
-   (1) happens before (2) because they are in the same thread (coordinator), and
-   (3) happens before (4) because they are in the same thread (worker).

If (2) has `Release` order, and (3) has `Acquire` order and (3) sees the value written by (2), then (2) happens before (3), too.  Then we have (1) HB (2) HB (3) HB (4).

Of course `SeqCst` is stronger than both `Acquire` and `Release` and has extra guarantees.  It is easiest to just use `SeqCst`.